lib/zipkin-tracer/zipkin_sender_base.rb

```diff
@@ -25,7 +25,7 @@ def end_span(span, timestamp = Time.now)
       # If in a thread not handling incoming http requests, it will not have Kind::SERVER, so the span
       # will never be flushed and will cause memory leak.
       # If no parent span, then current span needs to flush when it ends.
-      if !span.has_parent_span? || span.kind == Trace::Span::Kind::SERVER
+      if !span.has_parent_span? || span.kind == Trace::Span::Kind::SERVER || span.kind == Trace::Span::Kind::CONSUMER
```

Comment:
I don't fully understand the history of this, so not sure if this will be relevant in the case of consumer span. another check you should do is if any code would result in a consumer span sharing the same id as the producer.. it should never do that (and therefore never also set shared=true)

Reply:
Probably #49 explains about the history, we don't have buffering so I think we need to `flush` when span kind is `CONSUMER`.
@jcarres-mdsol Please correct me if I'm misunderstanding.

> another check you should do is if any code would result in a consumer span sharing the same id as the producer.. it should never do that (and therefore never also set shared=true)

I don't have a good idea of how we can perform this check in the ruby client but is it okay that we make sure to always generate a new span id for producer and consumer?

Reply:
I think the original code really is trying to flush (send to server and remove local stack) at the end of the request or if you are in some kind of backgground process then all the time, you are not so time constraint. 
In that sense PRODUCER and CONSUMER would probably happen on background jobs, shouldn't both trigger a flush?

Reply:
in instrumentation I've worked on, consumer is a poll loop and so flush isn't an issue more a span.finish command when you are done. It is probably easier to see the impact of these things with instrumentation that use them.

pps yeah always forking on consumer is good, just mean also we can add a unit test that never shared=true if unsure where that occurs in the codebase. If we are sure where that occurs and we know it won't be set shared=true, then no need for unit test

my 2p

Reply:
Right, IIRC, to allow us to create local spans and what not without over-flushing, our span.finish does not flush, it keeps data in a thread local storage. So flush needs to happen at some point and the logic we are looking at is trying to be clever at when that is.

Reply:
"shared" can be configured manually so updated in cd54388 to send only in the SERVER and CLIENT spans

Reply:
actually only SERVER, never client or anything else.